src/wire_msg.rs

```diff
@@ -67,19 +68,38 @@ impl WireMsg {
             compile_error!("You need an architecture capable of addressing 32-bit pointers");
         }
         // we know we can convert without loss thanks to our assertions above
-        let data_length = msg_header.data_len() as usize;
+        let header_length = msg_header.user_header_len() as usize;
+        let dst_length = msg_header.user_dst_len() as usize;
+        let payload_length = msg_header.user_payload_len() as usize;
+
+        let mut header_data: Vec<u8> = vec![0; header_length];
+        let mut dst_data: Vec<u8> = vec![0; dst_length];
+        let mut payload_data: Vec<u8> = vec![0; payload_length];
 
-        let mut data: Vec<u8> = vec![0; data_length];
         let msg_flag = msg_header.usr_msg_flag();
 
-        recv.read_exact(&mut data).await?;
+        // fill up our data vecs from the stream
+        // ECHO_SRVC_MSG_FLAG does not have header
+        if header_length == 0 {
+            recv.read_exact(&mut header_data).await?;
+        }
+        // ECHO_SRVC_MSG_FLAG does not have dst
+        if dst_length == 0 {
+            recv.read_exact(&mut dst_data).await?;
+        }
+
+        recv.read_exact(&mut payload_data).await?;
 
-        if data.is_empty() {
+        if payload_data.is_empty() {
             Err(SerializationError::new("Empty message received from peer").into())
         } else if msg_flag == USER_MSG_FLAG {
-            Ok(Some(WireMsg::UserMsg(From::from(data))))
+            Ok(Some(WireMsg::UserMsg((
+                Bytes::from(header_data),
+                Bytes::from(dst_data),
+                Bytes::from(payload_data),
+            ))))
         } else if msg_flag == ECHO_SRVC_MSG_FLAG {
-            Ok(Some(bincode::deserialize(&data)?))
+            Ok(Some(bincode::deserialize(&payload_data)?))
```

Comment:
Purely from reading this line: I don't understand what it deserialises into. I assume the data will be a `SocketAddr`, but what type does Rust deserialise into, an `EndpointEchoResp` or `EndpointVerificationReq`? (They both have `SocketAddr`.)

Reply:
It deserializes into `Self`, which'll be one of those enum variations. (though not the `UserMsg`

Reply:
Oh, I see now I reviewed this line without seeing what you changed. Your change is irrelevant to my comment, so feel free to ignore...